clang/test/SemaSYCL/builtin-fields.cpp

```diff
@@ -0,0 +1,230 @@
+  static_assert(is_same<decltype(__builtin_field_type(decltype(__builtin_field_type(C, 0)), 1)), int>::value, "expected an int");
+
+  // struct L has four fields despite containing an anonymous bit-field which
+  // is only sort of a field. All four fields are of type int despite some of
+  // the fields being bit-fields.
+  static_assert(__builtin_num_fields(L) == 4, "expected four fields");
+  static_assert(is_same<decltype(__builtin_field_type(L, 0)), int>::value, "expected an int");
+  static_assert(is_same<decltype(__builtin_field_type(L, 1)), int>::value, "expected an int");
+  static_assert(is_same<decltype(__builtin_field_type(L, 2)), int>::value, "expected an int");
+  static_assert(is_same<decltype(__builtin_field_type(L, 3)), int>::value, "expected an int");
+
+  // Virtual bases are still bases.
+  static_assert(__builtin_num_bases(M) == 1, "expected no bases");
+  static_assert(is_same<decltype(__builtin_base_type(M, 0)), A>::value, "expected an A");
+}
+
+template <typename Ty>
+void templates1() {
+  static_assert(__builtin_num_fields(F<Ty>) == 2, "expected two fields"); // expected-note {{in instantiation of template class 'F<S>' requested here}}
+  static_assert(__builtin_num_fields(G<2>) == 1, "expected one field");
+}
+
+template <typename Ty>
+void templates2() {
+  static_assert(__builtin_num_fields(Ty) == 3, "expected three fields");
+}
+
+template <typename FieldTy>
+void templates3() {
+  static_assert(__builtin_num_fields(F<FieldTy>) == 2, "expected two fields");
+  static_assert(is_same<decltype(__builtin_field_type(F<FieldTy>, 0)), FieldTy>::value, "expected a FieldTy");
+  static_assert(is_same<decltype(__builtin_field_type(F<FieldTy>, 1)), int>::value, "expected an int");
+}
+
+template <typename Ty>
+void templates4() {
+  static_assert(__builtin_num_bases(N<Ty>) == 3, "expected three bases");
+  static_assert(is_same<decltype(__builtin_base_type(N<Ty>, 0)), A>::value, "expected an A");
+  static_assert(is_same<decltype(__builtin_base_type(N<Ty>, 1)), D>::value, "expected a D");
+  static_assert(is_same<decltype(__builtin_base_type(N<Ty>, 2)), Ty>::value, "expected a Ty");
+}
+
+void instantiate() {
+  templates1<int>();
+  templates1<struct S>(); // expected-note {{in instantiation of function template specialization 'templates1<S>' requested here}} \
+                             expected-note {{forward declaration of 'S'}}
+  templates2<B>();
+  templates2<E>();
+
+  templates3<int>();
+  templates3<float>();
+  templates3<A>();
+
+  templates4<B>();
+}
+
+template <typename Func>
+void invoker(Func F) {
+  static_assert(__builtin_num_fields(Func) == 2, "expected one field");
+  static_assert(is_same<decltype(__builtin_field_type(Func, 0)), int>::value, "expected an int");
+  static_assert(is_same<decltype(__builtin_field_type(Func, 1)), double>::value, "expected a double");
+  F(1.0);

```

Comment:
What is this testing?

Reply:
`invoker()` is always called with a generic lambda, so this ensures the lambda gets instantiated with two different parameter types. This is important for the lambda on line 185 which has an `if constexpr` in it that looks at the parameter type.

I'll add comments.

Reply:
Thanks!